File: benchmarks/cpp/vector insertions and deletions/vector.cpp
#include <iostream>
#include <vector>
#include <chrono>

int main() {
    const size_t TOTAL = 10'000'000;   // scale this
    const size_t CHUNK = 10'000'000;      // 40 MB
    volatile long long sum = 0;

    auto start = std::chrono::high_resolution_clock::now();

    for (size_t done = 0; done < TOTAL; done += CHUNK) {
        size_t n = (TOTAL - done < CHUNK) ? (TOTAL - done) : CHUNK;

        std::vector<int> v;
        v.reserve(n);

        for (size_t i = 0; i < n; i++)
            v.push_back(i);

        while (!v.empty()) {
            sum += v.back();
            v.pop_back();
        }
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::cout << "C++ Vector time: "
              << std::chrono::duration<double>(end - start).count()
              << " s\n";
    std::cout << "Sum: " << sum << "\n";
}

File: benchmarks/rust/vector insertions and deletions/vector.rs
use std::time::Instant;

fn main() {
    let total: usize = 1_000_000_000; // scale this
    let chunk: usize = 10_000_000;    // 40 MB
    let mut sum: i64 = 0;

    let start = Instant::now();

    let mut done = 0;
    while done < total {
        let n = if total - done < chunk {
            total - done
        } else {
            chunk
        };

        let mut v: Vec<i32> = Vec::with_capacity(n);

        for i in 0..n {
            v.push(i as i32);
        }

        while let Some(x) = v.pop() {
            sum += x as i64;
        }

        done += n;
    }

    println!("Rust Vec time: {} s", start.elapsed().as_secs_f64());
    println!("Sum: {}", sum);
}
